Removal of ModSwitcher / Code cleanup

GhoulofGSG9 opened the pull request:
The development of the combat mod started almost 5 years ago.  At that point of time NS2 lacked any kind of support for mods. So many modders started writing their own mod frameworks. Combat still uses fsfod's Mod-Switcher as Framework to interact with the NS2 code base, However these days NS2 comes with the widely used ModLoader (entry file system) system which offers all features needed to implement combat. So it's time to switch from the previous used Mod Switcher to NS2's Modloader. 

The benefits of doing so is a massive reduction of the combat code base as a lot Mod-Switcher framework code isn't needed anymore.

Furthermore I removed all files that are not needed to play Combat. All in all this change should reduce the maintenance costs as much as possible.

To further simplify setting up file hooks I introduced the Filehooks directory:
Any file in the Post/Pre/Remove/Halt sub-directories will setup a Filehook of the type of the name of the subdirectory for a lua file with the same name and path in  the lua directory.

The current status of this PR can be checked at https://github.com/GhoulofGSG9/NS2Combat/projects/2 . I would recommend to test this version at the next SCC Mod play test before merging it. 

I completed all high priority tasks and will try to finish the normal priority task until the end of this week. The low priority tasks are unlikely to be completed by myself as they are beyond the scope of maintaining the combat mod.

Tasks:
- High Priority:
	- [x] Remove ModSwitcher/Switch to NS2's Modloader System 
	- [x] Fix the Alien buy menu (issue with backgrounds and selecting certain upgrades)
	- [x] First lua run-time optimization pass
- Normal Priority:
	- [x] Add a server config which allow server ops to tweak most of the global values
	- [x] Investigate and fix server crashes
	- [x] Balance pass (Make sure to keep values as close to Vanilla as possible)
	- [x] Fix the ink shader
	- [ ] Adjust bot logic to work with the combat buy menus
	- [ ] Make Combat compatible with NS2+
- Low Piority
	- [ ] Redesign the upgrade system to not use a techtree data structure (expensiove runtime and network bandwidth wise)
	- [ ] Optimize the used network model

GhoulofGSG9 committed:
Major code refactor:
- Removed all references to fsfod's modding framework and use the entry system
- Removed all not in use files and references (including Devour)
- Restructured the mod's folder structure

Todo: Fix the buy menus. Hide unneeded HUD elements

GhoulofGSG9 committed:
Fixed the alien buy menu

GhoulofGSG9 committed:
Fixed the marine buy menu

GhoulofGSG9 committed:
Fixed the alien buy menu upgrade button color schema
Fixed that you couldn't select the tier 2 and 3 upgrades at once

GhoulofGSG9 committed:
Merge remote-tracking branch 'nin/master' into feature/switch_to_entry_system

# Conflicts:
#	output/lua/ClassHooker.lua
#	output/lua/ClassUtility.lua
#	output/lua/Combat/Chat.lua
#	output/lua/Combat/CombatAlienUpgrade.lua
#	output/lua/Combat/CombatUpgrade.lua
#	output/lua/Combat/ExperienceData.lua
#	output/lua/Combat/ExperienceEnums.lua
#	output/lua/Combat/ExperienceFuncs.lua
#	output/lua/Combat/ExperienceLevels.lua
#	output/lua/Combat/FileHooks.lua
#	output/lua/Combat/FileHooks/Post/Alien.lua
#	output/lua/Combat/FileHooks/Post/Babbler.lua
#	output/lua/Combat/FileHooks/Post/ConsoleCommands_Client.lua
#	output/lua/Combat/FileHooks/Post/Effects.lua
#	output/lua/Combat/FileHooks/Post/GUIAlienBuyMenu.lua
#	output/lua/Combat/FileHooks/Post/Hydra.lua
#	output/lua/Combat/FileHooks/Post/MarineTeam.lua
#	output/lua/Combat/FileHooks/Post/Marine_Hooks.lua
#	output/lua/Combat/FileHooks/Post/NS2Utility.lua
#	output/lua/Combat/FileHooks/Post/NetworkMessages.lua
#	output/lua/Combat/FileHooks/Post/Player.lua
#	output/lua/Combat/FileHooks/Post/PlayingTeam.lua
#	output/lua/Combat/FileHooks/Post/ReadyRoomTeam.lua
#	output/lua/Combat/FileHooks/Post/SoundEffect.lua
#	output/lua/Combat/FileHooks/Post/TeamMessenger.lua
#	output/lua/Combat/FileHooks/Post/TechTree.lua
#	output/lua/Combat/FileHooks/Post/UpgradableMixin.lua
#	output/lua/Combat/FileHooks/Replace/GUIMarineBuyMenu.lua
#	output/lua/Combat/GUIGameTimeCountDown.lua
#	output/lua/Combat/Gobals.lua
#	output/lua/Combat/Locale.lua
#	output/lua/Combat/MarineBuyFuncs.lua
#	output/lua/Combat/Player_Upgrades.lua
#	output/lua/Combat/Player_Upgrades_Client.lua
#	output/lua/DevouredPlayer.lua
#	output/lua/DispatchBuilder.lua
#	output/lua/Hud/Alien/combat_GUIAlienSpectatorHUD.lua
#	output/lua/Hud/Alien/combat_GUIBioMassDisplay.lua
#	output/lua/Hud/Alien/combat_GUIUpgradeChamberDisplay.lua
#	output/lua/Hud/GUIDevourOnos.lua
#	output/lua/Hud/GUIDevouredPlayer.lua
#	output/lua/Hud/GUIFuncTrain.lua
#	output/lua/Hud/Marine/combat_GUIMarineHud.lua
#	output/lua/Hud/combat_GUIExperienceBar.lua
#	output/lua/Hud/combat_GUIPlayerResources.lua
#	output/lua/LoadTracker.lua
#	output/lua/ModUtility.lua
#	output/lua/PathUtil.lua
#	output/lua/Weapons/Alien/Devour.lua
#	output/lua/Weapons/Alien/combat_StructureAbility.lua
#	output/lua/Weapons/Marines/DevouredViewModel.lua
#	output/lua/Weapons/Marines/combat_ClipWeapon.lua
#	output/lua/combat_AlienTeam.lua
#	output/lua/combat_AlienTeamInfo.lua
#	output/lua/combat_AlienTeam_NewFuncs.lua
#	output/lua/combat_Armory.lua
#	output/lua/combat_Armory_Client.lua
#	output/lua/combat_Client.lua
#	output/lua/combat_CommandStation.lua
#	output/lua/combat_CommandStructure_Hooks.lua
#	output/lua/combat_ConsoleCommands.lua
#	output/lua/combat_DSPEffects.lua
#	output/lua/combat_DisorientableMixin.lua
#	output/lua/combat_Embryo.lua
#	output/lua/combat_FileOverrides.lua
#	output/lua/combat_Hive.lua
#	output/lua/combat_LiveMixin.lua
#	output/lua/combat_MenuModLauncher.lua
#	output/lua/combat_ModSwitcher.lua
#	output/lua/combat_NS2Gamerules.lua
#	output/lua/combat_NS2Gamerules_Hooks.lua
#	output/lua/combat_Onos.lua
#	output/lua/combat_PlayerHooks.lua
#	output/lua/combat_Player_ClientHook.lua
#	output/lua/combat_PowerPoint.lua
#	output/lua/combat_Predict.lua
#	output/lua/combat_Props.lua
#	output/lua/combat_Server.lua
#	output/lua/combat_Shared.lua
#	output/lua/combat_Team.lua
#	output/lua/combat_TechTreeHooks.lua
#	output/lua/combat_VersionsCheck.lua
#	output/lua/combat_Weapon.lua
#	output/lua/fsfod_scripts.lua

GhoulofGSG9 committed:
Optimize code to not use pairs where it's not needed

GhoulofGSG9 committed:
Fixed that the GUIMarineHUD filehook wasn't loaded

GhoulofGSG9 committed:
Removed method that already exists in Player_Upgrades.lua

GhoulofGSG9 committed:
Added Nin's small fixes from commit  bf2baf7 because they got lost during the last merge

GhoulofGSG9 committed:
Simplify the NS2Gamerules methods. Now the contain nothing more than the combat logic.

GhoulofGSG9 committed:
Removed all unnecessarily asset files.
Fixed that the shade cloud looked odd.

GhoulofGSG9 committed:
Fixed that the Marine hooks did not load

TheRealNin commented:
I think I'll publish a beta combat mod with these changes so they can be tested. This looks awesome.

GhoulofGSG9 commented:
I still need to figure out the odd consistency check issues that start occurring with this version mounted at the SCC server.

GhoulofGSG9 committed:
Fixed issues with getting the round time. No longer networking the round length and using the available vanilla round start time.
Fixed various smaller code style issues (unused vars/methods and vars using wrong scope)

GhoulofGSG9 committed:
Fixed the path to the Filehook folder which caused the mod to not load correctly on linux due to the file system being case sensitive.

GhoulofGSG9 committed:
Fixed a number of script error occurring because we tried to hook server side only functions at the client as well

GhoulofGSG9 committed:
Moved all server only functions from Player.lua to Player_Server.lua
Fixed that the taunt abilities did not work with custom taunts
Fixed that players spawned instantly instead after a short time.
Fixed that the waiting to spawn message did not show up for marines
Removed all usage of unpack
Optimized some of the PlayerUI functions

GhoulofGSG9 committed:
Added a server config file
Fixed that the chat commands didn't work

GhoulofGSG9 committed:
Fixed that the combat medpack behaved differently than the vanilla one

GhoulofGSG9 committed:
Fixed that the localization system didn't really work.
Created a new Balance.lua file just for vanilla balance tweaks

GhoulofGSG9 committed:
Fixed that the round time limit wasn't set correctly due to a typo

GhoulofGSG9 committed:
Don't even attempt to fix combat's tech system, without a rewrite this won't get any better.

Reverted all changes of  the previous commits which tried to work around the resend all upgrades at local client init situation.

GhoulofGSG9 committed:
Block commander bots from getting added.

GhoulofGSG9 committed:
Fixed that certain combat maps did not work because the mappers didn't assign the tech point correctly (either use 1/2 or 0/0 but not 0/1 or 0/2)

GhoulofGSG9 committed:
Fixed that the "press e to log in" dialog showed up near command structures

GhoulofGSG9 committed:
Fixed that the round time GUI was overlapped by the first person spectator GUI

GhoulofGSG9 committed:
Fixed that players never got removed from the re spawn queue.

GhoulofGSG9 committed:
Fixed that the overtime server config setting wasn't applied correctly.
Fixed that the reload speed upgrade didn't work
Foxed that armor and weapons upgrades did not show up correctly in the UI

GhoulofGSG9 committed:
Fixed a general typo in a player class method name

GhoulofGSG9 committed:
Merge remote-tracking branch 'nin/master' into feature/switch_to_entry_system

Additional changes:
Fixed the buy menu icons of the Rail Gun Exo, GL and MG
Fixed that you could buy a GL or FT after a MG

# Conflicts:
#	output/lua/Combat/CombatMarineUpgrade.lua
#	output/lua/Combat/ExperienceData.lua
#	output/lua/Combat/ExperienceEnums.lua
#	output/lua/Combat/FileHooks/Replace/GUIMarineBuyMenu.lua
#	output/lua/combat_MarineBuyFuncs.lua
#	output/lua/combat_NS2Gamerules_Hooks.lua
#	output/lua/combat_PlayingTeam.lua

GhoulofGSG9 committed:
Cleaned up the code comments

GhoulofGSG9 committed:
Fixed the predict script errors that occurred when you started to weld a command structure

GhoulofGSG9 commented:
After today's play test I think this PR is ready to be reviewed/merged. There are still some tasks that need to be finished. But those shouldn't stop this PR from getting merged.

GhoulofGSG9 committed:
Added all missing alien upgrades. Fixed some art style issues of the alien buy menu. It looks now more similar to the vanilla one.

GhoulofGSG9 committed:
Fixed some script errors that occurred at co_portals

GhoulofGSG9 committed:
Fixed various script errors (e.g. gorges healing the hive) which occurred due to changes being loaded in lua vm they aren't meant to load in.
